## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [47]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [48]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [49]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [50]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [51]:
# What three item categories had the most entries?
df.category.value_counts(ascending=False).head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [52]:
# For the category with the most entries, which subcategory had the most entries?
subcategory_df = df[df["category"] == "consumables"]
subcategory_df.subcategory.value_counts().idxmax()

'bathroom supplies'

In [53]:
# Which five clients had the most entries in the data?

df.client_id.value_counts(ascending=False).head()


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [54]:
# Store the client ids of those top 5 clients in a list.
client_list = df.client_id.value_counts(ascending=False).head(5).index.tolist()
client_list

[33615, 66037, 46820, 38378, 24741]

In [55]:
# How many total units (the qty column) did the client with the most entries order order?
total_units = df.loc[df['client_id'] == 33615, 'qty'].sum()
total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [56]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"] * df["qty"]
subtotal_df = df[["unit_price", "qty", "subtotal"]]
subtotal_df.head()

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [57]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def price(total_weight):
    if total_weight > 50:
        return 7
    else:
        return 10

    
df["shipping_price"] = df.apply(lambda row: row["unit_weight"] * row["qty"] * price(row["unit_weight"] * row["qty"]), axis=1)
shipping_price_df = df[["unit_price", "unit_weight", "qty",	"shipping_price"]]
shipping_price_df.head()


unit_price  unit_weight  qty  shipping_price
0     1096.80         7.50  105         5512.50
1       24.95         1.49   21          312.90
2       13.52         1.68   39          458.64
3       36.42         1.23   29          356.70
4      195.10        46.43   20         6500.20

In [58]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = df.apply(lambda row: ((row["subtotal"] + row["shipping_price"]) * .0925 + row["subtotal"] + row["shipping_price"]), axis=1).round(2)
total_price_df = df[["subtotal", "shipping_price", "total_price"]]
total_price_df.head()


subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40

In [59]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_line_cost"] = df.apply(lambda row: ((row["unit_cost"] * row["qty"]) + row["shipping_price"]), axis=1)
total_line_cost_df = df[["unit_cost", "qty", "shipping_price", "total_line_cost"]]
total_line_cost_df.head()

unit_cost  qty  shipping_price  total_line_cost
0     762.71  105         5512.50         85597.05
1      15.09   21          312.90           629.79
2       7.86   39          458.64           765.18
3      24.85   29          356.70          1077.35
4     108.17   20         6500.20          8663.60

In [60]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df.apply(lambda row: (row["total_price"] - row["total_line_cost"]), axis = 1)
profit_df = df[["total_price", "total_line_cost", "line_profit"]]
profit_df.head()

total_price  total_line_cost  line_profit
0    131839.08         85597.05     46242.03
1       914.26           629.79       284.47
2      1077.12           765.18       311.94
3      1543.57          1077.35       466.22
4     11364.40          8663.60      2700.80

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [61]:
# Check your work using the totals above

order2742071 = df.loc[df['order_id'] == 2742071, 'total_price'].sum()
print(f"Order ID 2742071 had a total price of: ${order2742071:.2f}")

order2173913 = df.loc[df['order_id'] == 2173913, 'total_price'].sum()
print(f"Order ID 2173913 had a total price of: ${order2173913:.2f}")

order6128929 = df.loc[df['order_id'] == 6128929, 'total_price'].sum()
print(f"Order ID 6128929 had a total price of: ${order6128929:.2f}")

Order ID 2742071 had a total price of: $152811.89
Order ID 2173913 had a total price of: $162388.71
Order ID 6128929 had a total price of: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [62]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

top5_df = df.groupby('client_id')['qty'].sum().reset_index().sort_values('qty', ascending=False).head(5)
top5_df

client_id      qty
758      78965  4033302
850      86721  1010426
204      30143   846069
840      85931   644423
738      77354   506382

In [63]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
columns_to_sum = ['qty', 'shipping_price', 'total_price', 'line_profit']
top5_summary_df = df.groupby('client_id')[columns_to_sum].sum().reset_index().sort_values('qty', ascending=False).head(5)
top5_summary_df = top5_summary_df.reset_index(drop=True)
top5_summary_df.rename(columns={"qty": "Total Units Purchased", "shipping_price" : "Total Shipping Price",
                       "total_price" : "Total Revenue", "line_profit" : "Total Profit", "client_id" : "Client_ID"},
                         inplace=True)
top5_summary_df

Client_ID  Total Units Purchased  Total Shipping Price  Total Revenue  \
0      78965                4033302            5340756.88   6.559237e+08   
1      86721                1010426           64335727.72   1.812663e+08   
2      30143                 846069            8981905.24   3.332885e+07   
3      85931                 644423           10122549.19   2.374170e+08   
4      77354                 506382           23814685.86   7.194265e+07   

   Total Profit  
0  2.473528e+08  
1  4.297470e+07  
2  1.125039e+07  
3  1.092941e+08  
4  1.243226e+07

In [64]:
# Format the data and rename the columns to names suitable for presentation.


# Define the money columns. 
money_columns = ["Total Shipping Price", "Total Revenue", "Total Profit"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    millions = amount / 1000000
    return millions

# Apply the currency_format_millions function to only the money columns. 
top5_summary_df[money_columns] = top5_summary_df[money_columns].apply(currency_format_millions, axis=1)

# Rename the columns to reflect the change in the money format. 
top5_summary_df.rename(columns={"Total Shipping Price" : "Total Shipping Price (millions)",
                                "Total Revenue" : "Total Revenue (millions)", 
                                "Total Profit" : "Total Profit (millions)"},
                                inplace=True)

top5_summary_df

Client_ID  Total Units Purchased  Total Shipping Price (millions)  \
0      78965                4033302                         5.340757   
1      86721                1010426                        64.335728   
2      30143                 846069                         8.981905   
3      85931                 644423                        10.122549   
4      77354                 506382                        23.814686   

   Total Revenue (millions)  Total Profit (millions)  
0                655.923670               247.352842  
1                181.266297                42.974695  
2                 33.328850                11.250386  
3                237.417001               109.294129  
4                 71.942652                12.432260

In [66]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_df = top5_summary_df.sort_values(by= "Total Profit (millions)", ascending=False)
final_df

Client_ID  Total Units Purchased  Total Shipping Price (millions)  \
0      78965                4033302                         5.340757   
3      85931                 644423                        10.122549   
1      86721                1010426                        64.335728   
4      77354                 506382                        23.814686   
2      30143                 846069                         8.981905   

   Total Revenue (millions)  Total Profit (millions)  
0                655.923670               247.352842  
3                237.417001               109.294129  
1                181.266297                42.974695  
4                 71.942652                12.432260  
2                 33.328850                11.250386